<?xml version="1.0" encoding="UTF-8"?>

<launch>

<arg name="localization"                  default="false"/>
<arg name="save_map"                      default="true"/>
<arg name="map_name"                      default="/home/xavier/maps_rtabmap/map.db"/>
<arg name="3d"                            default="true"/>
<arg name="start_rviz"                    default="true"/>
<arg unless="$(arg localization)"         name="args" default="--delete_db_on_start --Optimizer/GravitySigma 0.3"/>
<arg if="$(arg localization)"             name="args" default="--Optimizer/GravitySigma 0.3"/>

<!-- custom - select odom just from wheels (default) or from robot localization -->
<arg name="use_robot_local"             default="false"/>
<arg unless="$(arg use_robot_local)"    name="odom_topic"  default="/odom_wheels"/>
<!-- if use_robot_local is true, set the odom topic from the output of robot local -->
<arg if="$(arg use_robot_local)"        name="odom_topic"  default="/localization/odom_filtered"/>

<group ns="rtabmap">

    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg args)">
          
        <param name="frame_id" type="string" value="base_link"/>
        <param name="subscribe_depth" type="bool" value="false"/>
        <param name="subscribe_rgb" type="bool" value="false"/>
        <param name="subscribe_rgbd" type="bool" value="true"/>
        <param name="subscribe_scan" type="bool" value="true"/>
        <param name="queue_size" type="int" value="50"/>
          
        <remap from="odom" to="$(arg odom_topic)"/>
        <remap from="scan" to="/scan_filter"/>
        <remap from="rgbd_image" to="rgbd_image"/>
          
        <!-- RTAB-Map's parameters -->
        <param name="RGBD/ProximityPathMaxNeighbors"    type="string"   value="10"/>
        <param name="RGBD/NeighborLinkRefining"         type="string"   value="true"/>
        <param name="RGBD/ProximityBySpace"             type="string"   value="true"/>
        <param name="RGBD/AngularUpdate"                type="string"   value="0.01"/>
        <param name="RGBD/LinearUpdate"                 type="string"   value="0.01"/>
        <param name="RGBD/OptimizeFromGraphEnd"         type="string"   value="false"/>
        <param name="Reg/Force3DoF"                     type="string"   value="true"/>          
        <param name="Reg/Strategy"                      type="string"   value="0"/>       <!-- 0=Vis, 1=ICP (standard), 2=VisICP -->
          
        <!-- ICP parameters -->
        <param name="Icp/VoxelSize"                     type="string"   value="0.05"/>          
        <param name="Icp/MaxCorrespondenceDistance"     type="string"   value="0.05"/>    <!-- Standard was 0.1 -->
    
        <!-- Added Params -->
        <param name="approx_sync"                       type="bool"     value="true"/>
        <param name="gen_scan"                          type="bool"     value="true"/>
        <param name="map_always_update"                 type="bool"     value="false"/>   <!-- if true refresh grid map even if we are not moving-->
        <param name="map_empty_ray_tracing"             type="bool"     value="false"/>   <!-- if true fill empty space between the generated scans-->
        
        <!-- Parameters tuning for 3D mapping-->
	    <param name="Grid/3D" 						    type="string"   value="$(arg 3d)"/>
        <param name="Grid/FromDepth"     				type="string"   value="true"/>
	    <param name="Grid/MaxObstacleHeight" 			type="string"   value="2" />
	    <param name="Grid/RayTracing" 				    type="string"   value="true"/>
          
        <!-- Localization mode -->
        <param     if="$(arg localization)"     name="Mem/IncrementalMemory" type="string" value="false"/>
        <param unless="$(arg localization)"     name="Mem/IncrementalMemory" type="string" value="true"/>
        <param name="Mem/InitWMWithAllNodes"            type="string"   value="$(arg localization)"/>
        <param name="Rtabmap/StartNewMapOnLoopClosure"  type="string"   value="$(arg localization)"/>

        <!-- DB path -->
        <param unless="$(arg save_map)"       name="database_path"  type="string" value="~/.ros/rtabmap.db"/>
        <param     if="$(arg save_map)"       name="database_path"  type="string" value="$(arg map_name)"/>
        <param     if="$(arg localization)"   name="database_path"  type="string" value="$(arg map_name)"/>

        <!-- Reducing noise
	    <param name="cloud_noise_filtering_radius"          value="0.2"/>
	    <param name="cloud_noise_filtering_min_neighbors"   value="6"/>
	    -->
	  
	    <!-- Update Rate 
	    <param name="Rtabmap/DetectionRate"     value="1.2"/>
        -->
    </node>

</group>
  
<!-- Visualization RVIZ -->
<group if="$(arg start_rviz)">
    <param name="robot_description" command="$(find xacro)/xacro $(find state_machine)/urdf/lightrobot.urdf" />
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find state_machine)/config/rviz_conf.rviz"/>
</group>

</launch>
